snapd,snapctl: decode json using Number

stolowski opened the pull request:
Use json Decoder to avoid issues with big integers mentioned here: https://forum.snapcraft.io/t/snap-set-digits-as-string/1099 .

chipaca commented on client/client.go:
```diff
@@ -256,7 +258,9 @@ func (client *Client) doSync(method, path string, query url.Values, headers map[
 
 	if v != nil {
-		if err := json.Unmarshal(rsp.Result, v); err != nil {
+		dec := json.NewDecoder(strings.NewReader(string(rsp.Result)))
```
the `string` here forces a spurious copy of `rsp.Result`. I think you should use a `bytes.NewReader` here instead.

You've got a lot of them; all instances of `strings.NewReader(string(thing)` where the `thing` is a `[]byte` (o a `json.RawMessage`, same thing) should be a `bytes.NewReader(thing)` directly. As far as I can tell that's all instances of `strings.Newreader(string(thing))` in this code.

stolowski replied:
Thanks, you're totally right. Fixed.

mvo5 commented:
A bit of a meta-question: would it make sense to have a helper like `util/json.go:UnmarshalWithNumber()`   - with a better name etc. This would make the diff smaller and we would only have a single place that we need to change if we ever need to tweak this again. Same for NewDecoder (would it make sense to also move this into a helper that sets UserNumber()) right away?

stolowski commented:
@mvo5 as discussed in the HO, let's not do the helper at this point, we can do it later if there is a compelling reason.

niemeyer commented on client/client.go:
```diff
@@ -227,6 +227,7 @@ func (client *Client) do(method, path string, query url.Values, headers map[stri
 
 	if v != nil {
 		dec := json.NewDecoder(rsp.Body)
+		dec.UseNumber()
```
Can we have a unit test that breaks without those changes?  Otherwise this is likely to be dropped at some point and we won't notice.

niemeyer commented on cmd/snap/cmd_set.go:
```diff
@@ -70,8 +70,10 @@ func (x *cmdSet) Execute(args []string) error {
+		dec := json.NewDecoder(strings.NewReader(parts[1]))
+		dec.UseNumber()
+		err := dec.Decode(&value)
+		if err == nil && !dec.More() {
```
Same.

niemeyer commented on daemon/api.go:
```diff
@@ -1552,6 +1552,7 @@ func setSnapConf(c *Command, r *http.Request, user *auth.UserState) Response {
 
 	var patchValues map[string]interface{}
 	decoder := json.NewDecoder(r.Body)
+	decoder.UseNumber()
```
Same.

niemeyer commented on overlord/configstate/config/helpers.go:
```diff
@@ -56,7 +57,9 @@ func PatchConfig(snapName string, subkeys []string, pos int, config interface{},
-		err := json.Unmarshal([]byte(*config), &configm)
+		dec := json.NewDecoder(bytes.NewReader(*config))
+		dec.UseNumber()
+		err := dec.Decode(&configm)
```
Same, and so forth for the rest of these changes.

niemeyer commented:
I think the helper would be appropriate here, and perhaps avoid a little the need for so many unit tests. Also, the helper would do the More bit which is currently lacking in many of those cases. That may be a good solution to the above review question.

stolowski commented:
@niemeyer Thanks for the review. I've pushed the decoding helper. Working on more tests, so it's not ready for re-review yet.

stolowski commented:
The branch is ready for re-review; I've reverted the changes to state, task, snapstate as they were not needed after all, also added the helper and improved unit tests and the spread test.

mvo5 commented on overlord/configstate/config/helpers.go:
```diff
@@ -98,7 +100,9 @@ func GetFromChange(snapName string, subkeys []string, pos int, config map[string
 			raw = jsonRaw(value)
 		}
-		err := json.Unmarshal([]byte(*raw), result)
+		dec := json.NewDecoder(bytes.NewReader(*raw))
```
pardon my ignorance please, but why is this not using util.DecodeJsonWithNumbers() ? Maybe a small comment with the reason behind this?

stolowski replied:
Totally valid comment. My oversight, thanks for spotting!

mvo5 commented on overlord/configstate/config/helpers.go:
```diff
@@ -112,7 +116,9 @@ func GetFromChange(snapName string, subkeys []string, pos int, config map[string
 			raw = jsonRaw(value)
 		}
-		err := json.Unmarshal([]byte(*raw), &configm)
+		dec := json.NewDecoder(bytes.NewReader(*raw))
```
Same question as above

stolowski replied:
My oversight again.. thanks!

mvo5 commented on overlord/configstate/config/transaction_test.go:
```diff
@@ -62,7 +64,9 @@ func (op setGetOp) args() map[string]interface{} {
 		kv := strings.SplitN(pair, "=", 2)
 		var v interface{}
-		err := json.Unmarshal([]byte(kv[1]), &v)
+		dec := json.NewDecoder(strings.NewReader(kv[1]))
```
Would it make sense to use the helper here as well?

stolowski replied:
Yes, fixed that too.

mvo5 commented on overlord/configstate/config/transaction_test.go:
```diff
@@ -62,7 +64,9 @@ func (op setGetOp) args() map[string]interface{} {
-		err := json.Unmarshal([]byte(kv[1]), &v)
+		dec := json.NewDecoder(strings.NewReader(kv[1]))
+		dec.UseNumber()
+		err := dec.Decode(&v)
```
Not related to the PR, but this could be written in a single line: `if err := dec.Decode(&v); err != nil {`

stolowski replied:
Done.

mvo5 commented on overlord/hookstate/ctlcmd/set.go:
```diff
@@ -154,7 +154,9 @@ func (s *setCommand) setInterfaceSetting(context *hookstate.Context, plugOrSlot
 
 		var value interface{}
-		err := json.Unmarshal([]byte(parts[1]), &value)
+		dec := json.NewDecoder(strings.NewReader(parts[1]))
```
Would it make sense to use the helper here? If not, maybe a small comment why not?

stolowski replied:
Yes, another oversight. Fixed.

mvo5 commented on util/json.go:
```diff
@@ -0,0 +1,40 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
```
I'm not sure about `util` as a package name, it seems very generic. Maybe `jsonutil` ? I know it only carries a single helper for now, but I am concerned that it will become a bit of a grab-all package if we scoope it so wide.

stolowski replied:
Fair point. I agree, let's not let it become "grab-all package". Changed to jsonutil.

stolowski commented on client/client.go:
```diff
@@ -256,7 +257,8 @@ func (client *Client) doSync(method, path string, query url.Values, headers map[
 
 	if v != nil {
-		if err := json.Unmarshal(rsp.Result, v); err != nil {
+		err := util.DecodeJsonWithNumbers(bytes.NewReader(rsp.Result), v)
```
Done.

niemeyer commented on jsonutil/json.go:
```diff
@@ -0,0 +1,40 @@
+	"io"
+)
+
+// DecodeJsonWithNumbers decodes input data using json.Decoder, ensuring numbers are preserved
```
This seems better as _DecodeWithNumber,_ the "Json" part is implied as this is inside jsonutil, and the "Number" here refers to the type name, json.Number, rather than to "numbers" in general.

niemeyer commented on jsonutil/json.go:
```diff
@@ -0,0 +1,40 @@
+		return err
+	}
+	if dec.More() {
+		return fmt.Errorf("excess data found, not a valid json")
```
People won't associate "excess data" as the reason why, say, 15:45 was not accepted (which excess data?).  We ourselves were confused by that when debugging the problem.

I suggest simply saying "cannot parse json value".
